"""
–¢–µ—Å—Ç–æ–≤—ã–π gRPC –∫–ª–∏–µ–Ω—Ç –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ —Ä–∞–±–æ—Ç—ã –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–∞ —É–∫–ª–∞–¥–∫–∏ –≥—Ä–∞—Ñ–∞.
"""

import sys
from pathlib import Path

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ src
sys.path.insert(0, str(Path(__file__).parent / "src"))

import grpc
from generated import layout_pb2, layout_pb2_grpc


def test_grpc_service(host: str = 'localhost', port: int = 50051):
    """–¢–µ—Å—Ç–∏—Ä—É–µ—Ç gRPC —Å–µ—Ä–≤–∏—Å —Å —Ç–µ—Å—Ç–æ–≤—ã–º–∏ –¥–∞–Ω–Ω—ã–º–∏ –∏–∑ km_hand_layout.py"""
    
    print("üîß –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ gRPC —Å–µ—Ä–≤–∏—Å–∞ —É–∫–ª–∞–¥–∫–∏ –≥—Ä–∞—Ñ–∞...")
    
    # –¢–µ—Å—Ç–æ–≤—ã–µ –¥–∞–Ω–Ω—ã–µ –∏–∑ km_hand_layout.py
    test_links = [
        (0, 11), (0, 16), (0, 7), (1, 19), (1, 17), (1, 9), (1, 7), 
        (2, 8), (2, 11), (2, 18), (2, 19), (2, 9), (2, 12), (3, 10), 
        (3, 18), (3, 19), (4, 19), (5, 6), (5, 7), (5, 9), (6, 14), 
        (6, 9), (7, 8), (7, 10), (7, 9), (7, 12), (8, 15), (8, 10), 
        (9, 16), (9, 10), (12, 18), (12, 19), (13, 14), (15, 19), 
        (15, 17), (16, 13), (17, 12), (17, 11), (19, 16)
    ]
    
    test_blocks = list(range(20))  # –ë–ª–æ–∫–∏ 0-19
    
    try:
        # –°–æ–∑–¥–∞–µ–º gRPC –∫–∞–Ω–∞–ª
        channel = grpc.insecure_channel(f'{host}:{port}')
        stub = layout_pb2_grpc.LayoutServiceStub(channel)
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º –∑–¥–æ—Ä–æ–≤—å–µ —Å–µ—Ä–≤–∏—Å–∞
        print("üè• –ü—Ä–æ–≤–µ—Ä–∫–∞ –∑–¥–æ—Ä–æ–≤—å—è —Å–µ—Ä–≤–∏—Å–∞...")
        health_request = layout_pb2.HealthCheckRequest(service="layout")
        health_response = stub.HealthCheck(health_request)
        
        if health_response.status != layout_pb2.HealthCheckResponse.SERVING:
            print(f"‚ùå –°–µ—Ä–≤–∏—Å –Ω–µ –≥–æ—Ç–æ–≤: {health_response.message}")
            return False
        
        print("‚úÖ –°–µ—Ä–≤–∏—Å —Ä–∞–±–æ—Ç–∞–µ—Ç –Ω–æ—Ä–º–∞–ª—å–Ω–æ")
        
        # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å –Ω–∞ —É–∫–ª–∞–¥–∫—É
        request = layout_pb2.LayoutRequest()
        
        # –î–æ–±–∞–≤–ª—è–µ–º –±–ª–æ–∫–∏
        for block_id in test_blocks:
            block = request.blocks.add()
            block.id = str(block_id)
            block.content = f"–ë–ª–æ–∫ {block_id}"
            block.metadata["type"] = "test_block"
        
        # –î–æ–±–∞–≤–ª—è–µ–º —Å–≤—è–∑–∏
        for i, (source, target) in enumerate(test_links):
            link = request.links.add()
            link.id = f"link_{i}"
            link.source_id = str(source)
            link.target_id = str(target)
            link.metadata["type"] = "test_link"
        
        # –ù–∞—Å—Ç—Ä–∞–∏–≤–∞–µ–º –æ–ø—Ü–∏–∏
        request.options.sublevel_spacing = 150
        request.options.layer_spacing = 200
        request.options.optimize_layout = True
        
        # –û—Ç–ø—Ä–∞–≤–ª—è–µ–º –∑–∞–ø—Ä–æ—Å
        print(f"üìä –û—Ç–ø—Ä–∞–≤–∫–∞ –∑–∞–ø—Ä–æ—Å–∞: {len(request.blocks)} –±–ª–æ–∫–æ–≤, {len(request.links)} —Å–≤—è–∑–µ–π")
        response = stub.CalculateLayout(request)
        
        if response.success:
            print("‚úÖ –£–∫–ª–∞–¥–∫–∞ —É—Å–ø–µ—à–Ω–æ –∑–∞–≤–µ—Ä—à–µ–Ω–∞!")
            print(f"‚è±Ô∏è  –í—Ä–µ–º—è –≤—ã–ø–æ–ª–Ω–µ–Ω–∏—è: {response.statistics.processing_time_ms}–º—Å")
            print(f"üìè –†–∞–∑–º–µ—Ä—ã –≥—Ä–∞—Ñ–∞: {response.statistics.total_width:.1f} √ó {response.statistics.total_height:.1f} –ø–∏–∫—Å–µ–ª–µ–π")
            print(f"üèóÔ∏è –°—Ç—Ä—É–∫—Ç—É—Ä–∞: {response.statistics.total_levels} —É—Ä–æ–≤–Ω–µ–π, {response.statistics.total_sublevels} –ø–æ–¥—É—Ä–æ–≤–Ω–µ–π")
            print(f"üìà –ú–∞–∫—Å–∏–º–∞–ª—å–Ω—ã–π —Å–ª–æ–π: {response.statistics.max_layer}")
            print(f"üîó –ì—Ä–∞—Ñ –∞—Ü–∏–∫–ª–∏—á–µ—Å–∫–∏–π: {'–î–∞' if response.statistics.is_acyclic else '–ù–µ—Ç'}")
            print(f"üèùÔ∏è –ò–∑–æ–ª–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ –±–ª–æ–∫–∏: {response.statistics.isolated_blocks}")
            
            print("\nüìç –ü–µ—Ä–≤—ã–µ 5 –±–ª–æ–∫–æ–≤ —Å –ø–æ–∑–∏—Ü–∏—è–º–∏:")
            for block in response.blocks[:5]:
                print(f"  {block.id}: ({block.x:.0f}, {block.y:.0f}) "
                      f"—Å–ª–æ–π={block.layer}, —É—Ä–æ–≤–µ–Ω—å={block.level}, –ø–æ–¥—É—Ä–æ–≤–µ–Ω—å={block.sublevel_id}")
            
            return True
        else:
            print(f"‚ùå –û—à–∏–±–∫–∞ —É–∫–ª–∞–¥–∫–∏: {response.error_message}")
            return False
            
    except grpc.RpcError as e:
        print(f"‚ùå –û—à–∏–±–∫–∞ gRPC: {e.details() if hasattr(e, 'details') else str(e)}")
        return False
    except Exception as e:
        print(f"‚ùå –ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {str(e)}")
        return False


if __name__ == "__main__":
    print("üß™ –ó–∞–ø—É—Å–∫ —Ç–µ—Å—Ç–æ–≤ gRPC —Å–µ—Ä–≤–∏—Å–∞ —É–∫–ª–∞–¥–∫–∏ –≥—Ä–∞—Ñ–∞\n")
    
    # –ó–∞–ø—É—Å–∫–∞–µ–º —Ç–µ—Å—Ç
    success = test_grpc_service()
    
    print("\nüéâ –¢–µ—Å—Ç –∑–∞–≤–µ—Ä—à–µ–Ω!")
    
    if not success:
        print("\nüí° –ü–æ–¥—Å–∫–∞–∑–∫–∞: –£–±–µ–¥–∏—Ç–µ—Å—å —á—Ç–æ:")
        print("   1. –°–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω—ã protobuf —Ñ–∞–π–ª—ã: scripts\\generate_proto.bat")
        print("   2. –ó–∞–ø—É—â–µ–Ω —Å–µ—Ä–≤–µ—Ä: poetry run start-layout-service")
        print("   3. –°–µ—Ä–≤–µ—Ä –¥–æ—Å—Ç—É–ø–µ–Ω –ø–æ –∞–¥—Ä–µ—Å—É localhost:50051")